### PR TITLE
tools/makemanifest.py: Evaluate relative paths w.r.t. current file.

### PR DESCRIPTION
**Pull request by jimmo**
This makes it slightly easier to write manifests, especially when your manifest is out-of-tree.

Happy to undo the `MPY` -> `MPY_DIR` change. Mostly did this because to be consistent I would have used `BOARD`, but that would be different to `BOARD_DIR` in the Makefile. (And that's the first thing a user of this feature is likely to see -- in their mpconfigboard.mk, adding `FROZEN_MANIFEST = $(BOARD_DIR)/manifest.py`

Perhaps it should be `TOP` to match the Makefile instead of `MPY_DIR`?

I've tested this with an "out of tree" manifest that then `include('$(BOARD_DIR)/manifest.py')` which in turn `include('$(PORT_DIR)/manifest.py')`.

--

When loading a manifest file (e.g. by `include()`) it will chdir first.

Also add `$(MPY_LIB_DIR)`, `$(PORT_DIR)`, `$(BOARD_DIR)`, and rename
`$(MPY)` to `$(MPY_DIR)` to be consistent.

Updates existing manifests to match.

### CODE REVIEW

**Comment by stinos** [comment]
This got merged like yesterday and now there's a PR already, that's fast :P

Anyway: yes the relative paths are not ideal, but passing rather arbitrary positional arguments to a script also isn't: why not use argparse? Less error-prone, easier to extend later, easier to use since it produces a proper help text, and you can pass the result around instead of adding global variables.

File: tools/makemanifest.py [hunk 1]
```diff
@@ -115,7 +127,18 @@ def system(cmd):
         return -1, er.output
 
 def convert_path(path):
-    return path.replace('$(MPY)', TOP)
+    #Perform variable substituion. TODO: Other ENV vars?
```
Comment (stinos) [inline tools/makemanifest.py]: Linters can flag TODOs so the question is: is it really needed to have this here, or is this rather something which you wonder about yourself? Also: could be fixed by using argparse and having it parse arbitrary key-value pairs.

File: tools/makemanifest.py [hunk 2]
```diff
@@ -38,14 +38,22 @@ def include(manifest):
 
     The manifest argument can be a string (filename) or an iterable of
     strings.
+
+    Relative paths are resolved with respect to the current file.
```
Comment (stinos) [inline tools/makemanifest.py]: Is 'the current file' the `manifest` argument?

File: tools/makemanifest.py [hunk 3]
```diff
@@ -173,15 +196,18 @@ def freeze_internal(kind, path, script, opt):
         manifest_list.append((kind, path, script, opt))
 
 def main():
-    global TOP
+    global TOP, LIB, PORT, BOARD
```
Comment (stinos) [inline tools/makemanifest.py]: I'd use the same names as the replacements, easier to read.
But see previous comment: I still think using argparse and perhaps arbitrary key-value pars is a better solution overall

**Comment by robert-hh** [comment]
For the esp8266 file, the first line:
`freeze('modules')`
causes a compile error, because modules is not in the boards directory. Better write:
`freeze('$(MPY)/modules')`
Also: the coexistence of the old and new mechanism for to-be-frozen files is confusing. This lets me consider to let it mature for a while.

File: tools/makemanifest.py (hunk 2 above)
Comment (jimmo) [reply tools/makemanifest.py]: It's relative to the current file being executed. i.e. the file that called `include`. So it's not relative to the `manifest` argument, rather to the current file.

**Comment by jimmo** [comment]
Thanks @stinos - agreed about using argparse. I've made it use argparse and allow arbitrary vars. Still open to suggestions on whether these are the right names.

I've also added a warning about mpy-cross not being available c.f. #5218

**Comment by stinos** [comment]
Don;t really have an opinion on the names, I guess if previously MPY and LIB where fine, than MPY_DIR and MPY_LIB_DIR are already better.

File: tools/makemanifest.py [hunk 4]
```diff
@@ -173,23 +192,41 @@ def freeze_internal(kind, path, script, opt):
...
-    MPY_CROSS = TOP + '/mpy-cross/mpy-cross'
-    MPY_TOOL = TOP + '/tools/mpy-tool.py'
+    MAKE_FROZEN = VARS['MPY_DIR'] + '/tools/make-frozen.py'
+    MPY_CROSS = VARS['MPY_DIR'] + '/mpy-cross/mpy-cross'
+    MPY_TOOL = VARS['MPY_DIR'] + '/tools/mpy-tool.py'
+
+    # Ensure mpy-cross is built
+    if not os.path.exists(MPY_CROSS):
+        print('mpy-cross not found at {}'.format(MPY_CROSS))
+        print('Run `make` in {}'.format(os.path.dirname(MPY_CROSS)))
```
[the remark this answers is not in the document]
Comment (dpgeorge) [reply tools/makemanifest.py]: Agreed (eg one might be using gmake).  I changed this message to say "please build it first".

File: py/mkrules.mk [hunk 5]
```diff
@@ -100,7 +100,8 @@ $(HEADER_BUILD):
 ifneq ($(FROZEN_MANIFEST),)
 # to build frozen_content.c from a manifest
 $(BUILD)/frozen_content.c: FORCE $(BUILD)/genhdr/qstrdefs.generated.h
-	$(Q)$(MAKE_MANIFEST) -o $@ $(TOP) $(BUILD) "$(MPY_CROSS_FLAGS)" $(FROZEN_MANIFEST)
+	@$(ECHO) "GEN $@"
+	$(MAKE_MANIFEST) -o $@ -v "MPY_DIR=$(TOP)" -v "MPY_LIB_DIR=$(MPY_LIB_DIR)" -v "PORT_DIR=$(shell pwd)" -v "BOARD_DIR=$(BOARD_DIR)" -b "$(BUILD)" $(if $(MPY_CROSS_FLAGS),-f"$(MPY_CROSS_FLAGS)",) $(FROZEN_MANIFEST)
```
[the remark this answers is not in the document]
Comment (dpgeorge) [reply py/mkrules.mk]: It should be possible to use Python's `os.environ[...]` feature to get arbitrary environment variables, so probably not necessary to pass anything more here.

**Comment by dpgeorge** [comment]
Thanks for this!

I merged it in 8ba963cfa3da8e712e94c2429cfd6f064b1b4c69 with the following minor fixes/changes:
- fixed esp8266 and esp32 `freeze('modules')` to be `freeze('$(PORT_DIR)/modules')`
- put `$(Q)` back in front of the `$(MAKE_MANIFEST)` invocation
- put `echo GEN` back in `makemanifest.py` so it's printed *after* `echo MPY`

**Pull request closed without merging**